scripts: GCE helper

tbg opened the pull request:
This helper script was originally written by @petermattis and creates a GCE
instance which is useful for stress testing or other experiments which would
otherwise exceed what one wants to do on a personal computer.

Gave this script a general do-over. In particular, it applies @petermattis'
patch which allows for parallel builds, so that from a completely fresh repo:

```
$ time make build
real 2m11.776s
```

I experimented with local scratch SSDs, but these instances can only be
destroyed, not stopped, and the SSD must be manually formatted which
complicates the script to the point where it is more trouble than it's worth.

The instance is set up to shut down ~10 minutes after the users disconnects.
If that is not desired, `sudo touch /.active`. The idea here is to be
conservative by default since these machines run at around a dollar a minute.

I encourage everyone to try this - the setup of a new machine takes only a
few minutes and never again will your laptop melt.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8871)

<!-- Reviewable:end -->

tamird commented:
looks like stunnel wrapper got renamed in both commits?

---

Reviewed 2 of 3 files at r1, 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[scripts/alias.bash.gpr, line 3 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDsg5etTLxZ9HpsFcX:-KQDsg5ft5uPne_FSUJj:b-c6ykoq) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/alias.bash.gpr#L3)):*
> ```
> #!/usr/bin/env bash
> function gpr () {
>   hub fetch origin refs/pull/$1/head:pr/$1 && hub checkout pr/$1
> ```

there is also the current version which is fancier (and is a script, not an alias):
```
#!/usr/bin/env sh

set -eu

error_message() {
  ret=$?
  if [ $ret -ne 0 ]; then
    echo "Something went wrong. Did you push your branch? Did you set a remote tracking branch?"
  fi
  exit $ret
}

trap error_message EXIT

if [ $# == 0 ]; then
  branch_name=$(git rev-parse --abbrev-ref HEAD)
  remote_name=$(git config branch."$branch_name".remote)

  github_forked_org_name=$(git config remote."$remote_name".url | sed -E -e 's,^(https://|git@)github.com[:/]([^\./]+)/[^\./]+(\.git)?,\2,')
  github_origin_proj_url=$(git config remote.origin.url         | sed -E -e 's,^(https://|git@)github.com[:/]([^\./]+/[^\./]+)(\.git)?,\2,')

  url="https://github.com/$github_origin_proj_url/compare/master...$github_forked_org_name:$branch_name?expand="

  open "$url"
else
  hub fetch origin refs/pull/"$1"/head:pr/"$1" && hub checkout pr/"$1"
fi

```

---

*[scripts/bootstrap-debian.sh, line 6 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDtQOLd67-4ItFWGy3:-KQDtQOLd67-4ItFWGy4:bipp6w5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/bootstrap-debian.sh#L6)):*
> ```Shell
> # (with improved parallel build patches) and the cockroach repo.
> 
> set -eu
> ```

-o pipefail

---

*[scripts/bootstrap-debian.sh, line 10 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDsyPJi6HqXtIKMftL:-KQDsyPJi6HqXtIKMftM:b-xuvjal) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/bootstrap-debian.sh#L10)):*
> ```Shell
> GOVERSION="1.7"
> 
> sudo apt-get update -q && sudo apt-get install -q -y --no-install-recommends g++ gcc libc6-dev make git gdb patch bzip2
> ```

i think all but gdb and bzip2 are included in build-essential

---

*[scripts/bootstrap-debian.sh, line 21 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDtJ4nfmEAQGAb-LuK:-KQDtJ4nfmEAQGAb-LuL:b-bq7dl7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/bootstrap-debian.sh#L21)):*
> ```Shell
> 
> echo 'export GOPATH='"${HOME}"'; export PATH="'"${HOME}"'/go/bin:${GOPATH}/bin:${PATH}"' >> ~/.bashrc_go
> echo ". .bashrc_go" >> ~/.bashrc
> ```

nit: single quotes

---

*[scripts/gceslave.sh, line 1 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDtSajNd2m2YKBq2GI:-KQDtSajNd2m2YKBq2GJ:b-e9yd3p) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/gceslave.sh#L1)):*
> ```Shell
> #!/bin/bash
> ```

/usr/bin/env bash?

---

*[scripts/gceslave.sh, line 3 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDtS5ORvLjQ44EFVso:-KQDtS5ORvLjQ44EFVsp:bipp6w5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/gceslave.sh#L3)):*
> ```Shell
> #!/bin/bash
> 
> set -eu
> ```

-o pipefail

---

*[scripts/gceslave.sh, line 47 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDtjMEPswV2dXVBjyz:-KQDtjMEPswV2dXVBjz-:b-wl2mt0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/gceslave.sh#L47)):*
> ```Shell
>     ;;
>     *)
>     echo "$0: unknown command: $1"
> ```

it'd be good in general for this script to emit some help text

---

*[scripts/idle.sh, line 1 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDtoCq1Ul35H2z-4DO:-KQDtoCq1Ul35H2z-4DP:bdx5grs) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/idle.sh#L1)):*
> ```Shell
> #!/bin/bash
> ```

/usr/env/bin bash

set -euo pipefail

---

*[scripts/idle.sh, line 4 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDuAQZEuDxANOZiVmf:-KQDuAQZEuDxANOZiVmg:b-myszlc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/idle.sh#L4)):*
> ```Shell
> 
> # Crontab:
> # * * * * * /root/idle.sh 2>&1 >> /root/idle.log
> ```

is this used somewhere? perhaps the intention was to use it in gceslave?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-:-KQDsVuMsbRIvq58Bigx:b41weov)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Fixed.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[scripts/alias.bash.gpr, line 3 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDsg5etTLxZ9HpsFcX:-KQDwFhOuVt5-xyh77X_:b-dwz22n) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/alias.bash.gpr#L3)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> there is also the current version which is fancier (and is a script, not an alias):
> ```
> #!/usr/bin/env sh
> 
> set -eu
> 
> error_message() {
>   ret=$?
>   if [ $ret -ne 0 ]; then
>     echo "Something went wrong. Did you push your branch? Did you set a remote tracking branch?"
>   fi
>   exit $ret
> }
> 
> trap error_message EXIT
> 
> if [ $# == 0 ]; then
>   branch_name=$(git rev-parse --abbrev-ref HEAD)
>   remote_name=$(git config branch."$branch_name".remote)
> 
>   github_forked_org_name=$(git config remote."$remote_name".url | sed -E -e 's,^(https://|git@)github.com[:/]([^\./]+)/[^\./]+(\.git)?,\2,')
>   github_origin_proj_url=$(git config remote.origin.url         | sed -E -e 's,^(https://|git@)github.com[:/]([^\./]+/[^\./]+)(\.git)?,\2,')
> 
>   url="https://github.com/$github_origin_proj_url/compare/master...$github_forked_org_name:$branch_name?expand="
> 
>   open "$url"
> else
>   hub fetch origin refs/pull/"$1"/head:pr/"$1" && hub checkout pr/"$1"
> fi
> 
> ```
</details>
Whoops, didn't actually mean to commit this here. Mind opening a PR (at which point we should strip it from the wiki?)

---

*[scripts/bootstrap-debian.sh, line 6 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDtQOLd67-4ItFWGy3:-KQDwNZoLrcqQDt9ib_F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/bootstrap-debian.sh#L6)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> -o pipefail
</details>
Done.

---

*[scripts/bootstrap-debian.sh, line 10 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDsyPJi6HqXtIKMftL:-KQDwhbsQPhE7L2tdaoo:b-xkw5wa) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/bootstrap-debian.sh#L10)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i think all but gdb and bzip2 are included in build-essential
</details>
Trimmed the list.

---

*[scripts/bootstrap-debian.sh, line 21 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDtJ4nfmEAQGAb-LuK:-KQDwkT7ZLJmmkPSffEf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/bootstrap-debian.sh#L21)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: single quotes
</details>
Done.

---

*[scripts/gceslave.sh, line 1 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDtSajNd2m2YKBq2GI:-KQDxC9q-Yge0a7fSjz4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/gceslave.sh#L1)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> /usr/bin/env bash?
</details>
Done.

---

*[scripts/gceslave.sh, line 3 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDtS5ORvLjQ44EFVso:-KQDwo0OL6gWASvA5Ahs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/gceslave.sh#L3)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> -o pipefail
</details>
Done.

---

*[scripts/gceslave.sh, line 47 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDtjMEPswV2dXVBjyz:-KQDzHIuD4a8yyrd5Nq0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/gceslave.sh#L47)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> it'd be good in general for this script to emit some help text
</details>
Done.

---

*[scripts/idle.sh, line 4 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDuAQZEuDxANOZiVmf:-KQDxDphGUCZGtiwYG7K:b76p208) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/idle.sh#L4)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is this used somewhere? perhaps the intention was to use it in gceslave?
</details>
Unintentional add, removing. I didn't want another random script here for this; some of the logic is inlined in gceslave.sh.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-:-KQDxQbyL-cZz1J5ARm-:bt2vr1b)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 5 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[scripts/gceslave.sh, line 30 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQE0LZWR0PbM0Y85dbT:-KQE0LZWR0PbM0Y85dbU:b-xz51jz) ([raw file](https://github.com/cockroachdb/cockroach/blob/80e31264689a954659dbe670a375a4b29cc9898c/scripts/gceslave.sh#L30)):*
> ```Shell
>     # logging out for just a minute.
>     gcloud compute --project "${project}" ssh "${name}" --zone "${zone}" \
>     "echo 'bash -c \"w -hs | grep pts &> /dev/null || [[ -f /.active ]] || /sbin/shutdown -h now\"' | sudo crontab -"
> ```

why do you need `/sbin/`? why do you need `bash`?

---

*[scripts/alias.bash.gpr, line 3 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQDsg5etTLxZ9HpsFcX:-KQE06s4cHJp2Lg-8RE2:b23xr0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b3edbc3d3401371fbf794af940d9c8c1074b17f/scripts/alias.bash.gpr#L3)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Whoops, didn't actually mean to commit this here. Mind opening a PR (at which point we should strip it from the wiki?)
</details>
Yeah, will do.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[scripts/gceslave.sh, line 30 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQE0LZWR0PbM0Y85dbT:-KQE1dfxDOgtXfVn1_-O:beq7g81) ([raw file](https://github.com/cockroachdb/cockroach/blob/80e31264689a954659dbe670a375a4b29cc9898c/scripts/gceslave.sh#L30)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why do you need `/sbin/`? why do you need `bash`?
</details>
there's essentially nothing in ${PATH}. I had to play around a bit until this actually worked. Removed `bash` - not necessary after I made the thing POSIX (removing `&>`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r5.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-:-KQG2eMrYAQe5pa4C8fH:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR! I took the liberty of changing the merge base to `master` to make sure this script is available everywhere without extra cherry-picking.

tbg committed:
scripts: GCE helper

This helper script was originally written by @petermattis and creates a GCE
instance which is useful for stress testing or other experiments which would
otherwise exceed what one wants to do on a personal computer.

Gave this script a general do-over. In particular, it applies @petermattis'
patch which allows for parallel builds, so that from a completely fresh repo:

    $ time make build
    real 2m11.776s

I experimented with local scratch SSDs, but these instances can only be
destroyed, not stopped, and the SSD must be manually formatted which
complicates the script to the point where it is more trouble than it's worth.

The instance is set up to shut down ~10 minutes after the users disconnects.
If that is not desired, `sudo touch /.active`. The idea here (cribbed off a
snippet circulated by @dt) is to be conservative by default since these
machines run at around a dollar a minute.
The logic is smart enough to detect if a user logs back in after a shutdown
has been scheduled.

I encourage everyone to try this - the setup of a new machine takes only a
few minutes and never again will your laptop melt all over your lap.

tschottdorf commented:
Also had to tweak the shutdown mechanism more so that it would actually work reliably, but now it does (and it's better documented how).

bdarnell commented:
Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[scripts/gceslave.sh, line 13 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQIAyjpCp1PdAG4iO5R:-KQIAyjpCp1PdAG4iO5S:b-19cfuw) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d0f288fdddf3925fa8e0697dbec3b7e59ba32f3/scripts/gceslave.sh#L13)):*
> ```Shell
> case $1 in
>     create)
>     gcloud compute --project "cockroach-tschottdorf" \
> ```

This shouldn't be hard-coded, should it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8871)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Gah, how did that one slip through. I'll send a PR.

On Sun, Aug 28, 2016 at 6:07 PM Ben Darnell <notifications@github.com>
wrote:

> Review status: 1 of 3 files reviewed at latest revision, 1 unresolved
> discussion, all commit checks successful.
> ------------------------------
>
> *scripts/gceslave.sh, line 13 [r6]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/8871#-KQIAyjpCp1PdAG4iO5R:-KQIAyjpCp1PdAG4iO5S:b-19cfuw>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/6d0f288fdddf3925fa8e0697dbec3b7e59ba32f3/scripts/gceslave.sh#L13>):*
>
> case $1 in
>     create)
>     gcloud compute --project "cockroach-tschottdorf" \
>
> This shouldn't be hard-coded, should it?
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/8871>*
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/8871#issuecomment-243002513>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135Mre20j5KVIY7eu5gdAoapXk50sJks5qkgaugaJpZM4Ju04R>
> .
>
-- 

-- Tobias

RaduBerinde commented:
Awesome!!
